cpp/ql/src/Security/CWE/CWE-190/ComparisonWithWiderType.ql

```diff
@@ -49,7 +49,7 @@ where
   small = rel.getLesserOperand() and
   large = rel.getGreaterOperand() and
   rel = l.getCondition().getAChild*() and
-  upperBound(large).log2() > getComparisonSize(small) * 8 and
+  upperBound(large.getFullyConverted()).log2() > getComparisonSize(small) * 8 and
```

Comment:
I worry that this change could lead to false positives when the conversion is the other way around, widening rather than narrowing. If the range analysis can't determine that a `signed char x` is non-negative, then `(unsigned int)x` will have an upper bound of INT_MAX. I might be wrong, so please add a test to confirm.

To be on the safe side, you could _add_ the new line of code rather than replacing what was there.

Reply:
or
```
forall(Expr conv | conv = large.getConversion*() | upperBound(conv).log2() > getComparisonSize(small) * 8) and
```

Reply:
Yes, I have no principled argument for why the endpoints of the conversion should be special. But with both `forall` and `getConversion*` we'd have to check for performance regressions more carefully.

Reply:
I think I'm in favor of @geoffw0's more general solution as well. I'll add a problematic testcase first (the naive one I made didn't reveal any false positives, but I think I can tweak it so that it does the wrong thing), and see how the `forall` performs.